src/Compilers/Test/Core/CompilationVerifier.cs

```diff
@@ -228,6 +233,110 @@ public void VerifyTypeIL(string typeName, string expected)
             }
         }
 
+        private sealed class Resolver : ILVerify.ResolverBase
+        {
+            private readonly Dictionary<string, ImmutableArray<byte>> _imagesByName = new Dictionary<string, ImmutableArray<byte>>(StringComparer.OrdinalIgnoreCase);
+
+            internal Resolver(IList<ModuleData> allModuleData)
+            {
+                foreach (var module in allModuleData)
+                {
+                    string name = module.SimpleName;
+                    if (_imagesByName.ContainsKey(name))
+                    {
+                        throw new Exception($"Multiple modules named '{name}' were found");
+                    }
+                    _imagesByName.Add(name, module.Image);
+                }
+            }
+
+            protected override PEReader ResolveCore(string simpleName)
+            {
+                if (_imagesByName.TryGetValue(simpleName, out var image))
+                {
+                    return new PEReader(image);
+                }
+
+                throw new Exception($"ILVerify was not able to resolve a module named '{simpleName}'");
+            }
+        }
+
+        private void ILVerify(Verification verification)
+        {
+            if (verification == Verification.Skipped)
+            {
+                return;
+            }
+
+            try
+            {
+                var resolver = new Resolver(_allModuleData);
+                var verifier = new ILVerify.Verifier(resolver);
+
+                var mscorlibModule = _allModuleData.Single(m => m.IsCorLib);
+                verifier.SetSystemModuleName(new AssemblyName(mscorlibModule.SimpleName));
+
+                // Main module is the first one
+                var result = verifier.Verify(resolver.Resolve(_allModuleData[0].SimpleName));
+                if (result.Count() == 0)
+                {
+                    if ((verification & Verification.FailsILVerify) == 0)
+                    {
+                        return;
+                    }
+
+                    throw new Exception("IL Verify succeeded unexpectedly");
+                }
+
+                if ((verification & Verification.FailsILVerify) != 0)
+                {
+                    return;
+                }
+
+                string message = printVerificationResult(result);
+                throw new Exception("IL Verify failed unexpectedly: \r\n" + message);
+            }
+            catch (Exception)
```

Comment:
>Exception [](http://example.com/codeflow?start=19&length=9)

By catching all exceptions, aren't we potentially hiding bugs? For instance, `Microsoft.CodeAnalysis.VisualBasic.UnitTests.CodeGenTests.TestGetObjectValueCalls()` seems to fail `ILVerify()` because of a `NullReferenceException`.

Reply:
From previous discussion with Jared, we don't distinguish between the different failure modes of ILVerify. Either it succeeds or it fails. That's why I added a comment to the test:
```
            ' ILVerify null ref
            ' Tracked by https//github.com/dotnet/roslyn/issues/58652
```